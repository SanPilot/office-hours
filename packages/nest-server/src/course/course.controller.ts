import { Controller, Get, Param } from '@nestjs/common';
import { pick } from 'lodash';
import { Connection } from 'typeorm';
import { Course } from './course.entity';
import { GetCourseResponse } from '@template/common';
import { Question } from '../question/question.entity';

@Controller('courses')
export class CourseController {
  constructor(private connection: Connection) {}

  @Get(':id')
  async get(@Param() params): Promise<GetCourseResponse> {
    const course = await Course.findOne(params.id, {
      relations: ['officeHours', 'queues'],
    });
    return {
      name: course.name,
      officeHours: course.officeHours.map((e) =>
        pick(e, ['id', 'title', 'room', 'startTime', 'endTime']),
      ),
      queues: await Promise.all(
        course.queues.map(async (queue: any) => ({
          id: queue.id,
          room: queue.room,
          staffList: [
            //TODO get real data
            {
              id: 54,
              name: 'Will Stenzel',
              photoURL:
                'https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=24f4b12cccbf875c7740bbfed45a993900cf0d08d11aa07c84780b3a3501f3bacca4eb33ed5effee8aa2dd195750cfbc9884dd5f2ac62c8f',
            },
            {
              id: 42,
              name: 'Grisha Zaytsev',
              photoURL:
                'https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=471f2d695fbb8a00ee740ad3ea910453986aec81ddaecf889ae98b3a1858597b12650afd0d4e59c561172f76cb1946eec217ed89bd4074c0',
            },
          ],
          queueSize: await Question.count({ where: { queueId: queue.id } }),
        })),
      ),
    };
  }
}
