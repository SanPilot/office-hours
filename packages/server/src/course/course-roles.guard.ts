import { Injectable, CanActivate, ExecutionContext } from '@nestjs/common';
import { Reflector } from '@nestjs/core';
import { UserModel } from '../profile/user.entity';

@Injectable()
export class RolesGuard implements CanActivate {
  constructor(private reflector: Reflector) {}

  async canActivate(context: ExecutionContext): Promise<boolean> {
    const roles = this.reflector.get<string[]>('roles', context.getHandler());
    if (!roles) {
      return true;
    }
    const request = context.switchToHttp().getRequest();
    const user = await UserModel.findOne(request.user.userId, {
      relations: ['courses'],
    });

    const courseId = request.params.id;
    return this.matchRoles(roles, user, courseId);
  }

  matchRoles(roles: string[], user: UserModel, courseId: number): boolean {
    const userCourse = user.courses.find((course) => {
      return Number(course.courseId) === Number(courseId);
    });

    if (!userCourse) {
      return false;
    }

    const remaining = roles.filter((role) => {
      return userCourse.role.toString() === role;
    });

    return remaining.length > 0;
  }
}
