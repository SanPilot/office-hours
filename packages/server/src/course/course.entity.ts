import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  OneToMany,
  ManyToOne,
  JoinColumn,
} from 'typeorm';
import { OfficeHourModel } from './office-hour.entity';
import { QueueModel } from '../queue/queue.entity';
import { UserCourseModel } from '../profile/user-course.entity';
import { SemesterModel } from './semester.entity';
import { Exclude } from 'class-transformer';
import { Heatmap } from '@koh/common';

/**
 * Represents a course in the context of office hours.
 * @param id - The id number of this Course.
 * @param name - The subject and course number of this course. Ex: "CS 2500"
 * @param semester - The semester of this course.
 */
/*interface Course {
    id: number;
    name: string;
    url: string;
    semester: Semester;
    users: UserCourse[]
}*/

@Entity('course_model')
export class CourseModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @OneToMany((type) => OfficeHourModel, (oh) => oh.course)
  officeHours: OfficeHourModel[];

  @OneToMany((type) => QueueModel, (q) => q.course)
  queues: QueueModel[];

  @Column('text')
  name: string;

  @Column('text', { nullable: true })
  @Exclude()
  icalURL: string;

  @OneToMany((type) => UserCourseModel, (ucm) => ucm.course)
  @Exclude()
  userCourses: UserCourseModel;

  @ManyToOne((type) => SemesterModel, (semester) => semester.courses)
  @JoinColumn({ name: 'semesterId' })
  @Exclude()
  semester: SemesterModel;

  @Column({ nullable: true })
  @Exclude()
  // TODO: can we make these not nullable and work with TypeORM
  semesterId: number;

  @Column('boolean', { nullable: true })
  enabled: boolean; // Set to true if the given the course is using our app

  // The heatmap is false when there havent been any questions asked yet or there havent been any office hours
  heatmap: Heatmap | false;
}
