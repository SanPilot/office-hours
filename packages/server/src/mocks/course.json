{
  "schedule": [
    {
      "id": 84,
      "course_name": "CS 2500",
      "course_id": 169,
      "room": "RY 154",
      "ta_name": "Will Stenzel",
      "ta_photo": "https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=24f4b12cccbf875c7740bbfed45a993900cf0d08d11aa07c84780b3a3501f3bacca4eb33ed5effee8aa2dd195750cfbc9884dd5f2ac62c8f",
      "campus": 116,
      "semester": "Fall 2019",
      "queue": null,
      "start": "2019-09-21T12:00:00-04:00",
      "end": "2019-09-21T14:00:00-04:00",
      "swap": null,
      "canceled": false,
      "ta_arrived_on": null,
      "ta_departed_on": null
    },
    {
      "id": 82,
      "course_name": "CS 2500",
      "course_id": 169,
      "room": "KA 110",
      "ta_name": "Grigory Zaytsev",
      "ta_photo": "https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=471f2d695fbb8a00ee740ad3ea910453986aec81ddaecf889ae98b3a1858597b12650afd0d4e59c561172f76cb1946eec217ed89bd4074c0",
      "campus": 116,
      "semester": "Fall 2019",
      "queue": null,
      "start": "2019-09-04T06:00:00-04:00",
      "end": "2019-09-04T08:00:00-04:00",
      "swap": null,
      "canceled": false,
      "ta_arrived_on": null,
      "ta_departed_on": null
    }
  ],
  "queues": ["45"]
}
