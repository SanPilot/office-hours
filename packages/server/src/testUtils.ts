import { createConnection, Connection } from "../../ormconfig";
import { init } from "./server";
import path from "path";
import { Server } from "@hapi/hapi";

export function setupServerTest(): () => Server {
  let server: Server;

  beforeEach(async () => {
    server = await init();
  });

  afterEach(async () => {
    await server.stop();
  });

  return () => server;
}

export function setupDBTest() {
  let connection: Connection;
  beforeAll(async () => {
    connection = await createConnection({
      type: "postgres",
      url: "postgres://postgres@localhost:5432/test",
      synchronize: true,
      logging: false,
      entities: [path.resolve(__dirname, "entity/**/*{.js,.ts}")],
      migrations: [path.resolve(__dirname, "migration/**/*{.js,.ts}")],
      subscribers: [path.resolve(__dirname, "subscriber/**/*{.js,.ts}")],
    });
  });

  beforeEach(async () => {
    await connection.synchronize(true);
  });

  afterAll(async () => {
    await connection.close();
  });
}
