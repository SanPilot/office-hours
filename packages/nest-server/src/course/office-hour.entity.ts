import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  ManyToOne,
  JoinColumn,
  OneToMany,
} from "typeorm";
import { Course } from "./course.entity";

@Entity("office_hour")
export class OfficeHour extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => Course, (course) => course.officeHours)
  @JoinColumn({ name: "courseId" })
  course: Course;

  @Column({ nullable: true })
  courseId: number;

  @Column("text")
  title: string;

  @Column("text")
  room: string;

  @Column()
  startTime: Date;

  @Column()
  endTime: Date;
}
