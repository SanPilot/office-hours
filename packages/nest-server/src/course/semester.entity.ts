import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  OneToMany,
} from "typeorm";
import { Season } from "@template/common";
import { Course } from "./course.entity";

@Entity("semester_model")
export class Semester extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  season: Season;

  @Column()
  year: number;

  @OneToMany((type) => Course, (course) => course.semester)
  courses: Course[];
}
