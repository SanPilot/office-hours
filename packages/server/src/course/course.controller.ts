import {
  ClassSerializerInterceptor,
  Controller,
  Delete,
  Get,
  Param,
  Post,
  UseGuards,
  UseInterceptors,
  InternalServerErrorException,
} from '@nestjs/common';
import {
  GetCourseResponse,
  QueuePartial,
  Role,
  TACheckoutResponse,
} from '@koh/common';
import async from 'async';
import { Connection, getRepository, MoreThanOrEqual } from 'typeorm';
import { EventModel, EventType } from 'profile/event-model.entity';
import { JwtAuthGuard } from '../login/jwt-auth.guard';
import { Roles } from '../profile/roles.decorator';
import { User } from '../profile/user.decorator';
import { UserModel } from '../profile/user.entity';
import { QueueCleanService } from '../queue/queue-clean/queue-clean.service';
import { QueueModel } from '../queue/queue.entity';
import { CourseRolesGuard } from './course-roles.guard';
import { CourseModel } from './course.entity';
import { HeatmapService } from './heatmap.service';
import { OfficeHourModel } from './office-hour.entity';
import { QueueSSEService } from '../queue/queue-sse.service';
import moment = require('moment');

@Controller('courses')
@UseGuards(JwtAuthGuard, CourseRolesGuard)
@UseInterceptors(ClassSerializerInterceptor)
export class CourseController {
  constructor(
    private connection: Connection,
    private queueCleanService: QueueCleanService,
    private queueSSEService: QueueSSEService,
    private heatmapService: HeatmapService,
  ) {}

  @Get(':id')
  @Roles(Role.PROFESSOR, Role.STUDENT, Role.TA)
  async get(@Param('id') id: number): Promise<GetCourseResponse> {
    // TODO: for all course endpoint, check if they're a student or a TA
    const course = await CourseModel.findOne(id, {
      relations: ['queues', 'queues.staffList'],
    });

    // Use raw query for performance (avoid entity instantiation and serialization)
    course.officeHours = await getRepository(OfficeHourModel)
      .createQueryBuilder('oh')
      .select(['id', 'title', `"startTime"`, `"endTime"`])
      .where('oh.courseId = :courseId', { courseId: course.id })
      .getRawMany();
    course.heatmap = await this.heatmapService.getHeatmapFor(id);

    course.queues = await async.filter(
      course.queues,
      async (q) => await q.checkIsOpen(),
    );
    await async.each(course.queues, async (q) => {
      await q.addQueueTimes();
      await q.addQueueSize();
    });

    return course;
  }

  @Post(':id/ta_location/:room')
  @Roles(Role.PROFESSOR, Role.TA)
  async checkIn(
    @Param('id') courseId: number,
    @Param('room') room: string,
    @User() user: UserModel,
  ): Promise<QueuePartial> {
    let queue = await QueueModel.findOne(
      {
        room,
        courseId,
      },
      { relations: ['staffList'] },
    );

    if (!queue) {
      queue = await QueueModel.create({
        room,
        courseId,
        staffList: [],
        questions: [],
        allowQuestions: true,
      }).save();
    }

    if (queue.staffList.length === 0) {
      queue.allowQuestions = true;
    }

    queue.staffList.push(user);
    await queue.save();

    await EventModel.create({
      time: new Date(),
      eventType: EventType.TA_CHECKED_IN,
      user,
      courseId,
    }).save();

    await this.queueSSEService.updateQueue(queue.id);
    return queue;
  }

  @Delete(':id/ta_location/:room')
  @Roles(Role.PROFESSOR, Role.TA)
  async checkOut(
    @Param('id') courseId: number,
    @Param('room') room: string,
    @User() user: UserModel,
  ): Promise<TACheckoutResponse> {
    const queue = await QueueModel.findOne(
      {
        room,
        courseId,
      },
      { relations: ['staffList'] },
    );
    queue.staffList = queue.staffList.filter((e) => e.id !== user.id);
    if (queue.staffList.length === 0) {
      queue.allowQuestions = false;
    }
    await queue.save();

    await EventModel.create({
      time: new Date(),
      eventType: EventType.TA_CHECKED_OUT,
      user,
      courseId,
    }).save();

    const canClearQueue = await this.queueCleanService.shouldCleanQueue(queue);
    let nextOfficeHourTime = null;

    // find out how long until next office hour
    if (canClearQueue) {
      const soon = moment().add(15, 'minutes').toDate();
      const nextOfficeHour = await OfficeHourModel.findOne({
        where: { startTime: MoreThanOrEqual(soon) },
        order: {
          startTime: 'ASC',
        },
      });
      nextOfficeHourTime = nextOfficeHour?.startTime;
    }
    await this.queueSSEService.updateQueue(queue.id);
    return { queueId: queue.id, canClearQueue, nextOfficeHourTime };
  }
}
