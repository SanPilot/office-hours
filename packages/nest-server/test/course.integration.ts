import { setupIntegrationTest } from './util/testUtils';
import {CourseModule } from '../src/course/course.module';
import { CourseFactory, OfficeHourFactory } from './util/factories';

describe('Course Integraiton', () => {
  const supertest = setupIntegrationTest(CourseModule);

  describe('GET /courses/:id', () => {
    it('gets office hours and queues', async () => {
      const course = await CourseFactory.create({
        officeHours: [await OfficeHourFactory.create()],
      });
      const response = await supertest()
        .get(`/courses/${course.id}`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });
});
