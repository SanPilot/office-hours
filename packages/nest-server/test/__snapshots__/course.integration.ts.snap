// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseController (e2e) GET /courses/:id gets office hours and queues 1`] = `
Object {
  "name": "CS 2500",
  "officeHours": Array [
    Object {
      "endTime": "2020-05-20T15:30:00.000Z",
      "id": 1,
      "room": "WVH 101",
      "startTime": "2020-05-20T14:00:00.000Z",
      "title": "Alex & Stanley",
    },
  ],
  "queues": Array [],
}
`;
